javafmllanguage/src/main/java/net/minecraftforge/fml/javafmlmod/FMLModContainer.java

```
@@ -47,14 +52,79 @@ public FMLModContainer(IModInfo info, String className, ModFileScanData modFileS
+     * to the --add-exports and --add-opens command-line options:
+     *     Add-Exports: &lt;module&gt;/&lt;package&gt;( &lt;module&gt;/&lt;package&gt;)*
+     *     Add-Opens: &lt;module&gt;/&lt;package&gt;( &lt;module&gt;/&lt;package&gt;)*
+     *
+     * The value of each attribute is a space-separated list of slash-separated
+     * module-name/package-name pairs. A &gt;module&lt;/&gt;package&lt; pair in
+     * the value of an Add-Exports attribute has the same meaning as the
+     * command-line option --add-exports &gt;module&lt;/&gt;package&lt;=ALL-UNNAMED.
+     *
+     * A &gt;module&lt;/&gt;package&lt; pair in the value of an Add-Opens attribute has the
+     * same meaning as the command-line option --add-opens &gt;module&lt;/&gt;package&lt;=ALL-UNNAMED.
+     *
+     * Each attribute can occur at most once, in the main section of a MANIFEST.MF file.
+     * A particular pair can be listed more than once. If a specified module was not
+     * resolved, or if a specified package does not exist, then the corresponding pair
+     * is ignored.
+     */
+    private void openModules(ModuleLayer layer, Module self, SecureJar jar) throws NoSuchMethodException, SecurityException, IllegalAccessException, InvocationTargetException {
+        var manifest = jar.moduleDataProvider().getManifest();
+        var entry = manifest.getMainAttributes().getValue("Add-Opens");
+        if (entry != null)
+            addOpenOrExports(layer, self, true, entry);
+        entry = manifest.getMainAttributes().getValue("Add-Exports");
+        if (entry != null)
+            addOpenOrExports(layer, self, false, entry);
+    }
+
+    private void addOpenOrExports(ModuleLayer layer, Module self, boolean open, String entry) throws NoSuchMethodException, SecurityException, IllegalAccessException, InvocationTargetException {
+        for (var pair : entry.split(" ")) {
+            var pts = pair.trim().split("/");
+            if (pts.length == 2) {
```

There should be an else branch to throw on invalid Add-Exports/Add-Opens (e.g. length is 3) to aid debugging (rather than silently continuing without applying). The split method can also be changed to stop splitting after 3 slashes

---

I did this explicitly because it is required by the spec. And is how the [JVM itself](https://github.com/openjdk/jdk/blob/8b4749713c63a08e502845ed5d0a0236822018cd/src/java.base/share/classes/sun/launcher/LauncherHelper.java#L669) implements this method. I can add a branch to log a warning, but I wouldn't throw an exception.

---

Oh I see, in that case logging a warning would suffice, yeah